packages/runtime/src/runtime/hooks/useEnsembleAction.tsx

```diff
@@ -120,6 +131,163 @@ export const useInvokeApi: EnsembleActionHook<InvokeAPIAction> = (action) => {
+
+  useRegisterBindings(
+    {
+      files,
+    },
+    action?.id,
+    {
+      setFiles,
+    },
+  );
+
+  const inputEl = useMemo(() => {
+    const input = document.createElement("input");
+    input.type = "file";
+    input.multiple = action?.allowMultiple || false;
+    input.accept =
+      action?.allowedExtensions?.map((ext) => ".".concat(ext))?.toString() ||
+      "*/*";
+
+    input.onchange = (event: Event): void => {
+      const selectedFiles =
+        (event.target as HTMLInputElement).files || undefined;
+
+      if (selectedFiles) setFiles(selectedFiles);
+
+      if (onCompleteAction) onCompleteAction?.callback();
+    };
+
+    return input;
+  }, [action?.allowMultiple, action?.allowedExtensions, onCompleteAction]);
+
+  const callback = useCallback((): void => {
+    try {
+      inputEl.click();
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error(error);
+    }
+  }, [inputEl]);
+
+  return { callback };
+};
+
+export const useUploadFiles: EnsembleActionHook<UploadFilesAction> = (
+  action?: UploadFilesAction,
+) => {
+  const [body, setBody] = useState<Record<string, unknown>>();
+  const [headers, setHeaders] = useState<Record<string, unknown>>();
+  const [status, setStatus] = useState<UploadStatus>("pending");
+  const [progress, setProgress] = useState<number>(0.0);
+
+  const onCompleteAction = useEnsembleAction(action?.onComplete);
+  const onErrorAction = useEnsembleAction(action?.onError);
+  const screenContext = useScreenContext();
+
+  useRegisterBindings(
+    {
+      body,
+      headers,
+      status,
+      progress,
+    },
+    action?.id,
+    {
+      setBody,
+      setHeaders,
+      setStatus,
+      setProgress,
+    },
+  );
+
+  const apiModel = screenContext?.model?.apis?.find(
+    (model) => model?.name === action?.uploadApi,
+  );
+  if (!apiModel) return;
+
+  const progressCallback = (progressEvent: ProgressEvent): void => {
+    const percentCompleted = (progressEvent.loaded * 100) / progressEvent.total;
+
+    setProgress(percentCompleted);
+  };
+
+  const callback = async (): Promise<void> => {
+    const files = evaluate(
+      screenContext as ScreenContextDefinition,
+      action?.files,
+    ) as FileList | undefined;
+    if (!files || files.length === 0) throw Error("Files not found");
+
+    const formData = new FormData();
+    if (files.length === 1)
+      formData.append(action?.fieldName ?? "files", head(files) as Blob);
+    else
+      for (let i = 0; i < files.length; i++) {
+        formData.append(action?.fieldName ?? `file${i}`, files[i]);
+      }
+
+    const apiModelBody = apiModel?.body ?? {};
+    for (const key in apiModelBody) {
+      formData.append(key, apiModelBody[key] as string);
+    }
+    const resolvedInputs = Object.entries(action?.inputs ?? {})
+      .concat(
+        (apiModel?.inputs?.map((value) => [value, value]) as [
+          string,
+          unknown,
+        ][]) ?? [],
+      )
```

Comment:
Not sure why you're concatenating the declared inputs here? You should be checking the action.inputs keys against the declared inputs and only evaluating the intersection. All unspecified inputs default to undefined